vllm/v1/sample/rejection_sampler.py

```diff
@@ -288,25 +290,71 @@ def compute_probs(
     # NOTE(woosuk): `apply_top_k_top_p` uses sorting to calculate the mask,
     # which is slow for large vocab sizes. This may cause performance issues.
     logits = apply_top_k_top_p(logits, top_k, top_p)
+
+    if not sampling_metadata.no_penalties:
+        assert sampling_metadata.prompt_token_ids is not None
+
+        _, vocab_size = logits.shape
+
+        # Expand tensors from [batch_size] to [num_tokens].
+        prompt_token_ids = expand_batch_to_tokens(
+            sampling_metadata.prompt_token_ids,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+        output_tokens_t = _convert_to_tensors(
+            sampling_metadata.output_token_ids, vocab_size, logits.device)
+        output_tokens_t = expand_batch_to_tokens(
+            output_tokens_t,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+        presence_penalties = expand_batch_to_tokens(
+            sampling_metadata.presence_penalties,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+        frequency_penalties = expand_batch_to_tokens(
+            sampling_metadata.frequency_penalties,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+        repetition_penalties = expand_batch_to_tokens(
+            sampling_metadata.repetition_penalties,
+            cu_num_draft_tokens,
+            num_tokens,
+        )
+
+        # Apply penalties to logits.
+        logits = apply_penalties(
+            logits,
+            prompt_token_ids,
+            output_tokens_t,
+            presence_penalties,
+            frequency_penalties,
+            repetition_penalties,
+        )
+
     output_prob = logits.softmax(dim=-1, dtype=torch.float32)
     return output_prob
 
 
 def expand_batch_to_tokens(
-    x: torch.Tensor,  # [batch_size]
-    cu_num_tokens: torch.Tensor,  # [batch_size]
+    x: torch.Tensor,
+    cu_num_tokens: torch.Tensor,
     num_tokens: int,
     replace_from: int = 0,
     replace_to: int = 0,
 ) -> torch.Tensor:
-    """Expand [batch_size] tensor to [num_tokens] tensor based on the number of
-    tokens per batch in cu_num_tokens.
+    """Expand [batch_size] tensor to [num_tokens] tensor
+    (or [batch_size, vocab_size] to [num_tokens, vocab_size]) based on
```

Comment:
I actually don't understand in which case we have this 2D tensor. Could you please elaborate more?

Reply:
Is the second dim `vocab_size`? Shouldn't it be something like `seq_len`?